provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+import { EventEmitter } from 'events';
+import { errors, X509, SharedAccessSignature } from 'azure-iot-common';
+import * as machina from 'machina';
+import * as Provisioning from './transport_interface';
+import * as dbg from 'debug';
+const debug = dbg('azure-device-provisioning:transport-fsm');
+
+export interface TransportHandlers {
+  _doConnectForFsm(callback: (err?: Error) => void): void;
+  _doDisconnectForFsm(callback: (err?: Error) => void): void;
+  _doFirstRegistrationRequestForFsm(registrationId: string, authorization: SharedAccessSignature | X509 | string, requestBody: any, forceRegistration: boolean, callback: (err?: Error, responseBody?: any, result?: any, pollingInterval?: number) => void): void;
+  _doOperationStatusQueryForFsm(registrationId: string, operationId: string, callback: (err?: Error, responseBody?: any, result?: any, pollingInterval?: number) => void): void;
+  _getErrorFromResultForFsm(result: any): any;
+}
+
+export class TransportStateMachine extends EventEmitter implements Provisioning.Transport, TransportHandlers{
```

Comment:
>TransportStateMachine [](start = 13, length = 21)

This looks a lot like the beginning of the actual device client FSM.
 
A couple of things worry me:
1. this state machine doesn't have the appropriate abstraction level to deal with protocol specificities: MQTT errors, AMQP CBS links, etc. so it's likely you'll need another state machine in the lower transport layers

2. Looks like a case where inheritance and composition both work, but making it an inheritance thing makes it harder to test (which makes me lean in favor of composition). If we can get away with transports not having their own state machines, we could have separate "dumb" transport classes that implement the TransportHandlers contract and make this contract public. 

3. I don't know the details but polling seems inherently HTTP-ish. I would hope we don't have to poll in AMQP and MQTT and instead receive updates as messages over a link/topic. That would invalidate a lot of this FSM design.

Reply:
Agree on all counts.  Discussing in email.

---
In reply to: [146407891](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r146407891) [](ancestors = 146407891)